File: apps/ios/ExpCore/Sources/Domain/StartedRunMatch.swift
import Foundation

/// EXP-536: a remote start is only a COMMAND — the desktop inserts the
/// `coding_sessions` row a moment later — so every surface that sends one
/// waits for that row to sync in and then PUSHES the live session screen,
/// instead of parking a "watch it in the Agents tab" notice. These are the
/// matching rules, mirrored on Android (`domain/StartedRunMatch.kt`) and web
/// (`lib/started-run-match.ts`); a mismatch there is a start that never
/// navigates.
public enum StartedRunKey: Hashable, Sendable {
    /// An action run — keyed by the display-name SNAPSHOT, never the action
    /// id: the builtin "Create action" row carries `action_id` NULL.
    case action(name: String)
    /// A single-issue session — the desktop stamps `issue_id`.
    case issue(id: String)
    /// A batch run: `issue_id` NULL (it spans issues) and no action name.
    /// Nothing narrower exists — a batch row is unmatchable by issue
    /// server-side too — so the userId + startedAt cut carries the identity.
    case batch

    /// The key for a Start-coding send: 1 id is a plain session, 2+ a batch.
    public static func forIssues(_ issueIds: [String]) -> StartedRunKey? {
        guard let first = issueIds.first else { return nil }
        return issueIds.count >= 2 ? .batch : .issue(id: first)
    }
}

public enum StartedRunMatch {

    /// Clock-skew slack on the desktop-written `started_at` — a row stamped
    /// slightly before the send must still count as ours.
    public static let skew: TimeInterval = 120

    /// How long a start may take to surface its row before the watch gives
    /// up. The "waiting for the desktop" caption shares it: one that died
    /// first would strand a late push with no explanation, one that outlived
    /// it would spin forever.
    public static let deadline: TimeInterval = 180

    public static func matches(
        _ session: CodingSessionEntity,
        key: StartedRunKey,
        userId: String,
        cutoff: Date
    ) -> Bool {
        guard session.userId == userId else { return false }
        // Fail CLOSED on an unparseable stamp (unlike liveness): not pushing
        // beats pushing into somebody else's run.
        guard let startedAt = CodingSessionLiveness.parseIso(session.startedAt),
              startedAt >= cutoff
        else { return false }
        switch key {
        case let .action(name):
            return session.actionName == name
        case let .issue(id):
            return session.issueId == id && session.actionName == nil
        case .batch:
            return session.issueId == nil && session.actionName == nil
        }
    }

    /// The first row that IS this start, or nil while none has synced.
    public static func find(
        in sessions: [CodingSessionEntity],
        key: StartedRunKey,
        userId: String,
        cutoff: Date
    ) -> CodingSessionEntity? {
        sessions.first { matches($0, key: key, userId: userId, cutoff: cutoff) }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/StartedRunMatch.kt
package com.exponential.app.domain

import com.exponential.app.data.db.CodingSessionEntity
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.withTimeoutOrNull

// EXP-536: after ANY remote start the surface that sent it jumps into the
// live session screen instead of leaving a "watch it in Agents" chip behind.
// A start is only a COMMAND — the desktop inserts the coding_sessions row a
// moment later — so every caller waits for that row to sync in and matches it
// against what it just started. The matching rules are one shared object
// because all six Android start surfaces (Agents, Actions, issue list,
// issue detail, Reviews, Changes) must agree on them.

/** What a post-send session watch is looking for in the synced rows. */
sealed interface StartedRunKey {
    /** An action run — keyed by the display-name SNAPSHOT, never the action
     * id: the builtin "Create action" row carries `action_id` NULL. */
    data class Action(val actionName: String) : StartedRunKey

    /** A single-issue session — the desktop stamps `issue_id`. */
    data class Issue(val issueId: String) : StartedRunKey

    /** A batch run: `issue_id` NULL (it spans issues) and no action name.
     * Nothing narrower is available — a batch row is unmatchable by issue
     * server-side too — so the userId + startedAt cut carries the identity. */
    data object Batch : StartedRunKey

    companion object {
        /** The key for a Start-coding send: 1 id is a plain session, 2+ a batch. */
        fun forIssues(issueIds: List<String>): StartedRunKey? = when {
            issueIds.isEmpty() -> null
            issueIds.size >= 2 -> Batch
            else -> Issue(issueIds.first())
        }
    }
}

object StartedRunMatch {

    /** Clock-skew slack on the desktop-written `started_at` — a row stamped
     * slightly before the send must still count as ours. */
    const val SKEW_MS: Long = 120_000L

    /** How long a start may take to surface its row before the watch gives
     * up. Captions share it: one that died first would strand a late jump
     * with no explanation, one that outlived it would spin forever. */
    const val DEADLINE_MS: Long = 180_000L

    fun matches(
        session: CodingSessionEntity,
        key: StartedRunKey,
        userId: String,
        cutoffMs: Long,
    ): Boolean {
        if (session.userId != userId) return false
        // EXP-530: automation-started rows (started_reason non-null) are the
        // device's own doing — a user's pending start watch must never grab
        // one, even when the action name and timing line up.
        if (session.startedReason != null) return false
        if ((CodingSessionLiveness.parseEpochMs(session.startedAt) ?: 0L) < cutoffMs) return false
        return when (key) {
            is StartedRunKey.Action -> session.actionName == key.actionName
            is StartedRunKey.Issue -> session.issueId == key.issueId && session.actionName == null
            StartedRunKey.Batch -> session.issueId == null && session.actionName == null
        }
    }

    /**
     * Suspend until [sessions] carries the row for THIS start, or the
     * deadline passes (null). [sessions] is the caller's live coding_sessions
     * flow — already scoped to the account, so a switch mid-watch just stops
     * emitting.
     */
    suspend fun await(
        sessions: Flow<List<CodingSessionEntity>>,
        key: StartedRunKey,
        userId: String,
    ): String? {
        val cutoffMs = System.currentTimeMillis() - SKEW_MS
        return withTimeoutOrNull(DEADLINE_MS) {
            sessions.mapNotNull { rows ->
                rows.firstOrNull { matches(it, key, userId, cutoffMs) }
            }.first().id
        }
    }
}
